payroll-processor-functions/src/payroll-processor-functions/Features/Payrolls/PayrollTrigger.cs

```diff
@@ -0,0 +1,156 @@
+using Microsoft.Extensions.Logging;
+using Microsoft.WindowsAzure.Storage.Queue;
+using Microsoft.WindowsAzure.Storage.Table;
+using PayrollProcessor.Functions.Features.Employees;
+using PayrollProcessor.Functions.Infrastructure;
+using System;
+using System.Threading.Tasks;
+
+namespace PayrollProcessor.Functions.Features.Payrolls
+{
+    public class PayrollTrigger
+    {
+        [FunctionName("Payroll_GetAll")]
+        public async Task<IActionResult> GetPayrolls(
+            [HttpTrigger(AuthorizationLevel.Anonymous, "GET", Route = "payrolls")] HttpRequest req,
+            [Table("payrolls")] CloudTable payrollesTable,
+            ILogger log)
+        {
+            log.LogInformation($"Retrieving all payrolls: [{req}]");
+
+            var payrollQuerier = new TableQuerier(payrollesTable);
+
+            var payrolls = await payrollQuerier.GetAllData<Payroll, PayrollEntity>(e => PayrollEntity.Map.To(e));
+
+            return new OkObjectResult(Response.Generate(payrolls));
+        }
+
+        [FunctionName("EmployeePayrolls_GetAll")]
+        public async Task<IActionResult> GetAllPayrollsForEmployee(
+            [HttpTrigger(AuthorizationLevel.Anonymous, "GET", Route = "employees/{employeeId}/payrolls")] HttpRequest req,
+            [Table("employeePayrolls")] CloudTable employeePayrollsTable,
+            Guid employeeId,
+            ILogger log)
+        {
+            log.LogInformation($"Retrieving all payrolls for employee [{employeeId}]: [{req}]");
+
+            var payrollQuerier = new TableQuerier(employeePayrollsTable);
+
+            var payrolls = await payrollQuerier
+                .GetAllDataByPartitionKey<Payroll, EmployeePayrollEntity>(
+                    e => EmployeePayrollEntity.Map.To(e),
+                    employeeId.ToString("n"));
+
+            return new OkObjectResult(Response.Generate(payrolls));
+        }
+
+        [FunctionName("Payroll_Create")]
+        public async Task<IActionResult> CreatePayroll(
+                [HttpTrigger(AuthorizationLevel.Anonymous, "POST", Route = "payrolls")] HttpRequest req,
+                [Table("employees")] CloudTable employeeTable,
+                [Table("payrolls")] CloudTable payrollsTable,
+                [Queue("payroll-updates")] CloudQueue payrollUpdatesQueue,
+                ILogger log)
+        {
+            log.LogInformation($"Creating a new payroll: [{req}]");
+
+            var payroll = await Request.Parse<Payroll>(req);
+
+            payroll.Id = Guid.NewGuid();
+
+            var querier = new TableQuerier(employeeTable);
+
+            var option = await querier.GetEntity<EmployeeEntity, Employee>(
+                payroll.EmployeeDepartment.ToLowerInvariant(),
+                payroll.EmployeeId.ToString("n"),
+                EmployeeEntity.Map.To);
+
+            var employee = option.IfNone(() => throw new Exception($"Could not find employee [{payroll.EmployeeDepartment}] [{payroll.EmployeeId}]"));
```

Comment:
I found this `option.IfNone()` to be a nice way to exit early on a failure to get a required value that might not be found in the data persistence layer.

Reply:
much cleaner!